### PR TITLE
GAWB-1126 twitter typeahead this time

### PR DESCRIPTION
**Pull request by skwalker**
orch PR https://github.com/broadinstitute/firecloud-orchestration/pull/331
(old UI PR for reference https://github.com/broadinstitute/firecloud-ui/pull/577)

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [x] Tell the tech lead (TL) that the PR exists if they wants to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [x] **TL** sign off
- [ ] **LR** sign off
- [x] **Product Owner** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
     [org.broadinstitute.firecloud-ui.common :as common]
     [org.broadinstitute.firecloud-ui.common.components :as comps]
     [org.broadinstitute.firecloud-ui.common.modal :as modal]
     [org.broadinstitute.firecloud-ui.common.style :as style]
+    [org.broadinstitute.firecloud-ui.config :as config]
     [org.broadinstitute.firecloud-ui.endpoints :as endpoints]
     [org.broadinstitute.firecloud-ui.page.workspace.summary.library-utils :as library-utils]
-    [org.broadinstitute.firecloud-ui.utils :as utils]
-    ))
+    [org.broadinstitute.firecloud-ui.utils :as utils]    ))
```
Comment (rushtong) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs]: Move the closing `))` to the next line. In other words, revert this change.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs [hunk 2]
```diff
@@ -101,12 +102,13 @@
...
-          (fn [{:keys [property required inputHint renderHint]}]
+          (fn [{:keys [property required inputHint renderHint editable]}]
             (let [property-kwd (keyword property)
-                  {:keys [title type enum minimum consentCode]} (get-in library-schema [:properties property-kwd])
+                  {:keys [title type typeahead enum minimum consentCode]} (get-in library-schema [:properties property-kwd])
                   error? (contains? (:invalid-properties @state) property-kwd)
                   colorize (fn [style] (merge style (when error? {:borderColor (:exception-state style/colors)})))
-                  update-property #(swap! state update :attributes assoc property-kwd (.. % -target -value))]
+                  update-property #(swap! state update :attributes assoc property-kwd (.. % -target -value))
+                  disabled? (false? editable)]
```
Comment (rushtong) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs:112]: So ... I'm confused about what's happening here - I think your rebase is off. This might stomp on the LMSVN commits but I can't tell what git will do in the end. If I compare this branch's version against origin/dev, these lines are not changed.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs [hunk 3]
```diff
@@ -143,17 +145,55 @@
...
                                               :onChange update-property
                                               :rows 3})
+                     (= typeahead "ontology")
+                     (style/create-text-field {:ref property-kwd
+                                               :className "typeahead"
+                                               :placeholder "Select an ontology."
+                                               :style {:width "100%"}
+                                               :value (get (:attributes @state) property-kwd nil)
+                                               :onChange update-property
+                                               })
```
Comment (rushtong) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs]: move `})` to previous line

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs [hunk 4]
```diff
@@ -143,17 +145,55 @@
...
                      :else
                      (style/create-text-field {:style (colorize {:width "100%"})
                                                :type (cond (= renderHint "date") "date"
                                                            (= type "integer") "number"
                                                            :else "text")
                                                :min minimum
+                                               :disabled disabled?
```
Comment (rushtong) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs]: This is also a previous LMSVN commit.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs [hunk 5]
```diff
@@ -143,17 +145,55 @@
...
+                                   {:empty "<div> unable to find any matches to the current query </div>"
+                                    :suggestion
+                                    (fn [result]
+                                      (str "<div> <div style='font-weight: bold; line-height: 1.5em;'>" (aget result "label")
+                                           "</div><div>" (aget result "id") "</div>"
+                                           "<div style='font-size: small; font-style: italic'> " (aget result "definition") "</div></div>"))})}))
+       (.bind (js/$ property-kwd)
+              "typeahead:select"
+              (fn [ev suggestion]
+                (swap! state update :attributes assoc :library:diseaseOntologyLabel (aget suggestion "label") :library:diseaseOntologyID (aget suggestion "id"))))))})
```
Comment (rushtong) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs]: It would be nice if this function were not specific to catalog and instead put into endpoints or even it's own component file so we can re-use it. I think it could be more generic that way too ... pass in the search box binding, an autocomplete term, and an endpoint and let the function do all the work in returning content.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/library_view.cljs [hunk 6]
```diff
@@ -11,22 +11,25 @@
 
 
 (defn- render-value [value]
+  ;; figure out what to do with typeahead?
```
Comment (rushtong) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/library_view.cljs]: Is this needed?

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/library_view.cljs [hunk 7]
```diff
@@ -11,22 +11,25 @@
...
   (cond (sequential? value) (clojure.string/join ", " value)
         (common/attribute-list? value) (clojure.string/join ", " (common/attribute-values value))
         :else value))
 
-(defn- render-library-row [key value]
+(defn- render-library-row [key value doid] ;; doid
```
Comment (rushtong) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/library_view.cljs]: remove comment.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs [hunk 8]
```diff
@@ -143,17 +145,55 @@
                                               :value (get (:attributes @state) property-kwd)
                                               :onChange update-property
                                               :rows 3})
+                     (= typeahead "ontology")
+                     (style/create-text-field {:ref property-kwd
+                                               :className "typeahead"
+                                               :placeholder "Select an ontology."
```
Comment (davidangb) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs]: nit: this should be "select an ontology value" or "select a value". We're using the DOID ontology, so the end user can't actually choose a different ontology.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs [hunk 9]
```diff
@@ -143,17 +145,55 @@
...
+   :component-did-mount
+   (fn [{:keys [props state refs]}]
+     (let [options (js/Bloodhound.
+                     (clj->js
+                       {:datumTokenizer js/Bloodhound.tokenizers.whitespace
+                        :queryTokenizer js/Bloodhound.tokenizers.whitespace
+                        :remote (clj->js {:url (str (config/api-url-root) "/duos/autocomplete/%QUERY")
+                                          :wildcard "%QUERY"})}))
+           existing-ontology (get-in props [:workspace :workspace :library-attributes :library:diseaseOntologyLabel])
+           property-kwd (@refs "library:diseaseOntologyLabel")] ;; fix this
```
Comment (davidangb) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs]: `library:diseaseOntologyLabel` in the previous two lines should not be hardcoded; it needs to be passed in as an argument for any given property.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs (hunk 5 above)
Comment (davidangb) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs]: more hardcoded property names

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/library_view.cljs [hunk 10]
```diff
@@ -11,22 +11,25 @@
...
 
 
 (defn- render-property [library-schema library-attributes property-key]
   (render-library-row
    (get-in library-schema [:properties property-key :title])
-   (render-value (get library-attributes property-key))))
-
+   (render-value
+    (or (get library-attributes property-key) (get-in library-schema [:properties property-key :default])))
+    (if (= (get-in library-schema [:properties property-key :typeahead]) "ontology") (get library-attributes :library:diseaseOntologyID) nil)))
```
Comment (davidangb) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/library_view.cljs]: hardcoded property name

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspaces_list.cljs [hunk 11]
```diff
@@ -179,13 +188,17 @@
...
+                                (->> (merge preds (:selected-types @state))
+                                     (keep (fn [[k v]] (when (and (not (false? v)) (some? (preds k))) k)))
+                                     (map preds)
+                                     (cons (constantly false)) ;; keeps (apply some-fn) from bombing when the list is empty
+                                     (apply some-fn)))]
+                  (filter
+                    (every-pred
+                      (somepred access-predicates)
+                      (somepred realm-predicates))
+                    (:workspaces props)))
```
Comment (davidangb) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspaces_list.cljs:201]: I think this entire file is a bad rebase.

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs [hunk 12]
```diff
@@ -143,17 +145,55 @@
...
+                     {:source options
+                      :display (fn [result]
+                                 (aget result "label"))
+                      :templates (clj->js
+                                   {:empty "<div> unable to find any matches to the current query </div>"
+                                    :suggestion
+                                    (fn [result]
+                                      (str "<div> <div style='font-weight: bold; line-height: 1.5em;'>" (aget result "label")
+                                           "</div><div>" (aget result "id") "</div>"
+                                           "<div style='font-size: small; font-style: italic'> " (aget result "definition") "</div></div>"))})}))
```
Comment (rushtong) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs]: My language alarm bell goes off when I see raw html in anything other than an html/jsp/etc. Is there no way to pass a vector to a method that then builds the content using standard `[:div {}]` syntax?

**Review by davidangb — COMMENTED** [review]
comments inline:
* I'd love to see the hardcoded property names disappear (or create a follow-on story to clean them up)
* worried about the rebase
* functionally, this allows me to save an arbitrary string into the diseaseOntologyLabel field - I start typing, make sure I haven't selected anything from the typeahead, and save. Is this correct? (Need Moran's input here)

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/library_view.cljs [hunk 13]
```diff
@@ -11,22 +11,25 @@
...
 
 
 (defn- render-property [library-schema library-attributes property-key]
   (render-library-row
    (get-in library-schema [:properties property-key :title])
-   (render-value 
-    (or (get library-attributes property-key) (get-in library-schema [:properties property-key :default])))))
+   (render-value
+    (or (get library-attributes property-key) (get-in library-schema [:properties property-key :default])))
+    (if (= (get-in library-schema [:properties property-key :typeahead]) "ontology") (get library-attributes :library:diseaseOntologyID) nil)))
```
Comment (davidangb) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/library_view.cljs]: to avoid hardcoding, how about something like (pseudocode):
```
if (typeahead === "ontology") {
  replace a "Label" suffix with an "ID" suffix on the property-key and display the ID
}
```

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs (hunk 12 above)
Comment (skwalker) [reply src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs]: I asked @dmohs earlier and he said no, unfortunately..

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/library_view.cljs [hunk 14]
```diff
@@ -15,18 +15,20 @@
...
```
Comment (davidangb) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/library_view.cljs]: 1) un-indent this line - it should be parallel to the `(render-value` and `(get-in` preceding it
2) this still has a hardcoded `:library:diseaseOntologyID`. It should be ... guessing here ... `(get-in library-schema [:properties property-key : relatedID])`

**Comment by davidangb** [comment]
one comment inline about a straggler hardcoded property name. Also still need decision on whether or not this should allow saving an arbitrary value (see previous comment). Otherwise looking good!

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs [hunk 15]
```diff
@@ -154,8 +163,41 @@
...
+                                       :minLength 1})
+                             (clj->js
+                               {:source options
+                                :display (fn [result]
+                                           (aget result "label"))
+                                :templates (clj->js
+                                             {:empty "<div> unable to find any matches to the current query </div>"
+                                              :suggestion
+                                              (fn [result]
+                                                (str "<div> <div style='font-weight: bold; line-height: 1.5em;'>" (aget result "label")
```
Comment (davidangb) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs]: by making the label bold, the `highlight` feature of the typeahead library doesn't work for label - because it's trying to bold something that's already bold. Highlighting is cool - maybe this needs some CSS tweaks?

File: src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs [hunk 16]
```diff
@@ -144,6 +146,22 @@
...
                                               :onChange update-property
                                               :rows 3})
+                     (= typeahead "ontology")
+                     [:div {}
+                      (style/create-text-field {:ref property-kwd
+                                                :className "typeahead"
+                                                :placeholder "Select an ontology value."
+                                                :style {:width "100%" :marginBottom "0px"}})
+                      (let [relatedID (library-utils/get-related-id (:attributes @state) library-schema property-kwd
+                                                                      (library-utils/get-related-id (:attributes props) library-schema property-kwd nil))
```
Comment (davidangb) [inline src/cljs/org/broadinstitute/firecloud_ui/page/workspace/summary/catalog.cljs]: this feels a little convoluted ... it's looking for the related id in state, and falls back to props if it's not in state? Is this necessary? I'm not quite following the logic, so I leave it up to you.

**Review by davidangb — COMMENTED** [review]
one code comment inline.

I bet this could use some visual UI polish, but I'm good with pushing this PR in and coming back to clean up the visuals afterwards. Nice job!

**Review by davidangb — APPROVED** [review]

**Comment by davidangb** [comment]
👍 with latest commits.
